File: php/src/CnpjValidator.php
<?php

declare(strict_types=1);

namespace LibCnpj;

final class CnpjValidator
{
    private const LENGTH = 14;
    private const BASE_LENGTH = 12;
    private const ASCII_ZERO = 48;
    private const ALL_ZEROS = '00000000000000';

    /**
     * Weights for the first check digit (positions 1..12).
     */
    private const WEIGHTS_FIRST = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];

    /**
     * Weights for the second check digit (positions 1..13).
     */
    private const WEIGHTS_SECOND = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];

    public static function isValid(string $value): bool
    {
        if (CnpjFormatter::containsOnlyAllowedCharacters($value) === false) {
            return false;
        }

        $strippedValue = CnpjFormatter::strip($value);

        if (strlen($strippedValue) !== self::LENGTH) {
            return false;
        }

        if ($strippedValue === self::ALL_ZEROS) {
            return false;
        }

        $base = substr($strippedValue, 0, self::BASE_LENGTH);
        $informedCheckDigits = substr($strippedValue, self::BASE_LENGTH, 2);

        if (self::checkDigitsAreNumeric($informedCheckDigits) === false) {
            return false;
        }

        $calculatedCheckDigits = self::calculateCheckDigits($base);

        if ($calculatedCheckDigits === null) {
            return false;
        }

        if ($calculatedCheckDigits !== $informedCheckDigits) {
            return false;
        }

        return true;
    }

    public static function calculateCheckDigits(string $base): ?string
    {
        if (strlen($base) !== self::BASE_LENGTH) {
            return null;
        }

        if (self::containsOnlyAlphanumericDigits($base) === false) {
            return null;
        }

        $firstDigit = self::calculateDigit($base, self::WEIGHTS_FIRST);

        $baseWithFirstDigit = $base . $firstDigit;
        $secondDigit = self::calculateDigit($baseWithFirstDigit, self::WEIGHTS_SECOND);

        return $firstDigit . $secondDigit;
    }

    private static function containsOnlyAlphanumericDigits(string $base): bool
    {
        $length = strlen($base);

        for ($index = 0; $index < $length; $index = $index + 1) {
            if (self::isAlphanumericDigit($base[$index]) === false) {
                return false;
            }
        }

        return true;
    }

    private static function checkDigitsAreNumeric(string $digits): bool
    {
        $length = strlen($digits);

        for ($index = 0; $index < $length; $index = $index + 1) {
            if (self::isNumericDigit($digits[$index]) === false) {
                return false;
            }
        }

        return true;
    }

    private static function isAlphanumericDigit(string $character): bool
    {
        if ($character >= '0' && $character <= '9') {
            return true;
        }

        if ($character >= 'A' && $character <= 'Z') {
            return true;
        }

        return false;
    }

    private static function isNumericDigit(string $character): bool
    {
        if ($character >= '0' && $character <= '9') {
            return true;
        }

        return false;
    }

    /**
     * @param array<int, int> $weights
     */
    private static function calculateDigit(string $value, array $weights): string
    {
        $sum = 0;
        $length = strlen($value);

        for ($index = 0; $index < $length; $index = $index + 1) {
            $digitValue = ord($value[$index]) - self::ASCII_ZERO;
            $sum = $sum + ($digitValue * $weights[$index]);
        }

        $remainder = $sum % 11;

        if ($remainder < 2) {
            return '0';
        }

        return (string) (11 - $remainder);
    }
}

File: java/src/main/java/com/libcnpj/CnpjValidator.java
package com.libcnpj;

public final class CnpjValidator {

    private static final int LENGTH = 14;
    private static final int BASE_LENGTH = 12;
    private static final int ASCII_ZERO = 48;
    private static final String ALL_ZEROS = "00000000000000";

    private static final int[] WEIGHTS_FIRST = {5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};
    private static final int[] WEIGHTS_SECOND = {6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};

    private CnpjValidator() {
    }

    public static boolean isValid(String value) {
        if (CnpjFormatter.containsOnlyAllowedCharacters(value) == false) {
            return false;
        }

        String strippedValue = CnpjFormatter.strip(value);

        if (strippedValue.length() != LENGTH) {
            return false;
        }

        if (ALL_ZEROS.equals(strippedValue)) {
            return false;
        }

        String base = strippedValue.substring(0, BASE_LENGTH);
        String informedCheckDigits = strippedValue.substring(BASE_LENGTH, LENGTH);

        if (checkDigitsAreNumeric(informedCheckDigits) == false) {
            return false;
        }

        String calculatedCheckDigits = calculateCheckDigits(base);

        if (calculatedCheckDigits == null) {
            return false;
        }

        return calculatedCheckDigits.equals(informedCheckDigits);
    }

    public static String calculateCheckDigits(String base) {
        if (base.length() != BASE_LENGTH) {
            return null;
        }

        if (containsOnlyAlphanumericDigits(base) == false) {
            return null;
        }

        String firstDigit = calculateDigit(base, WEIGHTS_FIRST);
        String baseWithFirstDigit = base + firstDigit;
        String secondDigit = calculateDigit(baseWithFirstDigit, WEIGHTS_SECOND);

        return firstDigit + secondDigit;
    }

    private static boolean containsOnlyAlphanumericDigits(String base) {
        for (int index = 0; index < base.length(); index = index + 1) {
            if (isAlphanumericDigit(base.charAt(index)) == false) {
                return false;
            }
        }

        return true;
    }

    private static boolean checkDigitsAreNumeric(String digits) {
        for (int index = 0; index < digits.length(); index = index + 1) {
            if (isNumericDigit(digits.charAt(index)) == false) {
                return false;
            }
        }

        return true;
    }

    private static boolean isAlphanumericDigit(char character) {
        if (character >= '0' && character <= '9') {
            return true;
        }

        if (character >= 'A' && character <= 'Z') {
            return true;
        }

        return false;
    }

    private static boolean isNumericDigit(char character) {
        if (character >= '0' && character <= '9') {
            return true;
        }

        return false;
    }

    private static String calculateDigit(String value, int[] weights) {
        int sum = 0;

        for (int index = 0; index < value.length(); index = index + 1) {
            int digitValue = value.charAt(index) - ASCII_ZERO;
            sum = sum + (digitValue * weights[index]);
        }

        int remainder = sum % 11;

        if (remainder < 2) {
            return "0";
        }

        return Integer.toString(11 - remainder);
    }
}
